File: typescript/processor/src/dsti/intent-extractor.ts
/**
 * @docspec:module id="docspec-ts-intent-extractor" name="DSTI Intent Extractor"
 * @docspec:description "Extracts all 13 DSTI intent signal channels from TypeScript
 *   method AST nodes. Channels: nameSemantics, guardClauses, branches, dataFlow,
 *   loopProperties, errorHandling, constants, nullChecks, assertions, logStatements,
 *   dependencies, validationAnnotations, returnType (via nameSemantics). Uses the
 *   TypeScript Compiler API for precise syntax tree analysis."
 * @docspec:boundary "processor-internal"
 * @docspec:since "3.0.0"
 */
import ts from "typescript";

export interface ExtractedSignals {
  nameSemantics?: { verb?: string; object?: string; intent?: string };
  guardClauses?: number;
  branches?: number;
  dataFlow?: { reads: string[]; writes: string[] };
  loopProperties?: { hasStreams?: boolean; hasEnhancedFor?: boolean; streamOps?: string[] };
  errorHandling?: { catchBlocks: number; caughtTypes: string[] };
  constants?: string[];
  dependencies?: string[];
  nullChecks?: number;
  assertions?: number;
  validationAnnotations?: number;
  logStatements?: number;
  intentDensityScore?: number;
}

const VERB_INTENT_MAP: Record<string, string> = {
  get: "query", find: "query", fetch: "query", load: "query", read: "query", list: "query", search: "query",
  create: "creation", add: "creation", insert: "creation", save: "creation", register: "creation",
  update: "mutation", modify: "mutation", patch: "mutation", set: "mutation", change: "mutation",
  delete: "deletion", remove: "deletion", destroy: "deletion", drop: "deletion",
  validate: "validation", check: "validation", verify: "validation", assert: "validation",
  process: "transformation", handle: "transformation", execute: "transformation", transform: "transformation",
  convert: "transformation", map: "transformation", parse: "transformation",
  send: "communication", publish: "communication", emit: "communication", notify: "communication",
  calculate: "calculation", compute: "calculation", count: "calculation", sum: "calculation",
  is: "query", has: "query", can: "query",
};

export class IntentExtractor {
  extractFromMethod(node: ts.MethodDeclaration, sourceFile: ts.SourceFile): ExtractedSignals | null {
    const name = node.name?.getText(sourceFile);
    if (!name) return null;

    const signals: ExtractedSignals = {};

    // 1. Name semantics
    const words = name.replace(/([A-Z])/g, " $1").trim().split(/\s+/);
    const verb = words[0].toLowerCase();
    signals.nameSemantics = {
      verb,
      object: words.slice(1).join(""),
      intent: VERB_INTENT_MAP[verb] ?? "unknown",
    };

    // 2-13. Analyze method body
    if (node.body) {
      const bodyText = node.body.getText(sourceFile);
      signals.guardClauses = this.countGuards(node.body);
      signals.branches = this.countBranches(node.body);
      signals.nullChecks = this.countNullChecks(bodyText);
      signals.logStatements = this.countLogStatements(bodyText);
      signals.assertions = this.countAssertions(bodyText);
      signals.constants = this.extractConstants(node.body, sourceFile);
      signals.dependencies = this.extractDependencies(node.body, sourceFile);
      signals.errorHandling = this.analyzeErrorHandling(node.body, sourceFile);
      signals.loopProperties = this.analyzeLoops(node.body);
      signals.dataFlow = this.analyzeDataFlow(node.body, sourceFile);
      signals.validationAnnotations = this.countValidationAnnotations(bodyText);
    }

    return signals;
  }

  private countGuards(body: ts.Block): number {
    let count = 0;
    const visit = (node: ts.Node) => {
      if (ts.isIfStatement(node) && ts.isBlock(node.thenStatement)) {
        const stmts = node.thenStatement.statements;
        if (stmts.length === 1 && (ts.isThrowStatement(stmts[0]) || ts.isReturnStatement(stmts[0]))) {
          count++;
        }
      }
      ts.forEachChild(node, visit);
    };
    visit(body);
    return count;
  }

  private countBranches(body: ts.Block): number {
    let count = 0;
    const visit = (node: ts.Node) => {
      if (ts.isIfStatement(node) || ts.isConditionalExpression(node) || ts.isSwitchStatement(node)) count++;
      ts.forEachChild(node, visit);
    };
    visit(body);
    return count;
  }

  private countNullChecks(text: string): number {
    return (text.match(/!==?\s*(null|undefined)|===?\s*(null|undefined)|\?\./g) ?? []).length;
  }

  private countLogStatements(text: string): number {
    return (text.match(/console\.(log|warn|error|info|debug)|logger\./g) ?? []).length;
  }

  private countAssertions(text: string): number {
    return (text.match(/assert|expect\(|should\./g) ?? []).length;
  }

  private countValidationAnnotations(text: string): number {
    // class-validator type decorators
    const classValidatorTypes = (text.match(
      /@(IsString|IsEmail|IsNumber|IsInt|IsBoolean|IsDate|IsEnum|IsOptional|IsNotEmpty|IsArray|IsUUID|IsUrl)\b/g
    ) ?? []).length;

    // class-validator constraint decorators
    const classValidatorConstraints = (text.match(
      /@(Min|Max|MinLength|MaxLength|Length|Matches|Contains)\b/g
    ) ?? []).length;

    // class-validator advanced decorators
    const classValidatorAdvanced = (text.match(
      /@(ValidateNested|ValidateIf|IsDefined)\b/g
    ) ?? []).length;

    // Joi validation patterns (e.g., Joi.string(), Joi.number(), Joi.object(), etc.)
    const joiPatterns = (text.match(
      /Joi\.(string|number|boolean|date|object|array|alternatives|any|binary|func|link|ref|symbol)\s*\(/g
    ) ?? []).length;

    // Zod validation patterns (e.g., z.string(), z.number(), z.object(), etc.)
    const zodPatterns = (text.match(
      /z\.(string|number|boolean|date|object|array|enum|union|intersection|tuple|record|map|set|literal|nativeEnum|optional|nullable|any|unknown|void|never|undefined|null|bigint|nan|symbol|function|lazy|preprocess|pipeline|coerce)\s*\(/g
    ) ?? []).length;

    return classValidatorTypes + classValidatorConstraints + classValidatorAdvanced + joiPatterns + zodPatterns;
  }

  private extractConstants(body: ts.Block, sf: ts.SourceFile): string[] {
    const constants: string[] = [];
    const visit = (node: ts.Node) => {
      if (ts.isStringLiteral(node) && node.text.length > 1) constants.push(node.text);
      if (ts.isNumericLiteral(node)) constants.push(node.text);
      ts.forEachChild(node, visit);
    };
    visit(body);
    return [...new Set(constants)].slice(0, 10);
  }

  private extractDependencies(body: ts.Block, sf: ts.SourceFile): string[] {
    const deps: string[] = [];
    const visit = (node: ts.Node) => {
      if (ts.isPropertyAccessExpression(node) && ts.isIdentifier(node.expression)) {
        const obj = node.expression.text;
        if (obj.startsWith("this.") === false && obj !== "console" && obj !== "Math" && obj !== "JSON") {
          deps.push(`${obj}.${node.name.text}`);
        }
      }
      if (ts.isCallExpression(node) && ts.isPropertyAccessExpression(node.expression)) {
        const expr = node.expression;
        if (ts.isPropertyAccessExpression(expr) && (expr.expression.kind === ts.SyntaxKind.ThisKeyword)) {
          deps.push(expr.name.text);
        }
      }
      ts.forEachChild(node, visit);
    };
    visit(body);
    return [...new Set(deps)].slice(0, 20);
  }

  private analyzeErrorHandling(body: ts.Block, sf: ts.SourceFile): { catchBlocks: number; caughtTypes: string[] } {
    let catchBlocks = 0;
    const caughtTypes: string[] = [];
    const visit = (node: ts.Node) => {
      if (ts.isTryStatement(node) && node.catchClause) {
        catchBlocks++;
        if (node.catchClause.variableDeclaration?.type) {
          caughtTypes.push(node.catchClause.variableDeclaration.type.getText(sf));
        }
      }
      ts.forEachChild(node, visit);
    };
    visit(body);
    return { catchBlocks, caughtTypes };
  }

  private analyzeLoops(body: ts.Block): { hasStreams?: boolean; hasEnhancedFor?: boolean; streamOps?: string[] } {
    let hasStreams = false;
    let hasEnhancedFor = false;
    const streamOps: string[] = [];
    const streamMethods = ["map", "filter", "reduce", "forEach", "flatMap", "find", "some", "every", "sort"];
    const visit = (node: ts.Node) => {
      if (ts.isForOfStatement(node)) hasEnhancedFor = true;
      if (ts.isCallExpression(node) && ts.isPropertyAccessExpression(node.expression)) {
        const methodName = node.expression.name.text;
        if (streamMethods.includes(methodName)) {
          hasStreams = true;
          if (!streamOps.includes(methodName)) streamOps.push(methodName);
        }
      }
      ts.forEachChild(node, visit);
    };
    visit(body);
    return {
      hasStreams: hasStreams || undefined,
      hasEnhancedFor: hasEnhancedFor || undefined,
      streamOps: streamOps.length > 0 ? streamOps : undefined,
    };
  }

  private analyzeDataFlow(body: ts.Block, sf: ts.SourceFile): { reads: string[]; writes: string[] } {
    const reads: string[] = [];
    const writes: string[] = [];
    const visit = (node: ts.Node) => {
      if (ts.isBinaryExpression(node) && node.operatorToken.kind === ts.SyntaxKind.EqualsToken) {
        if (ts.isPropertyAccessExpression(node.left)) {
          writes.push(node.left.name.text);
        }
      }
      if (ts.isPropertyAccessExpression(node) && (node.expression.kind === ts.SyntaxKind.ThisKeyword)) {
        reads.push(node.name.text);
      }
      ts.forEachChild(node, visit);
    };
    visit(body);
    return { reads: [...new Set(reads)], writes: [...new Set(writes)] };
  }
}

File: rust/processor/src/dsti/intent_extractor.rs
//! Extract intent signals from Rust AST using syn.
//!
//! @docspec:module {
//!   id: "docspec-rust-dsti-extractor",
//!   name: "Intent Signal Extractor",
//!   description: "Extracts DSTI intent signals from Rust function signatures and bodies using syn-based AST analysis and pattern matching across 13 channels.",
//!   since: "3.0.0"
//! }

use serde_json::Value;
use std::collections::HashMap;
use std::sync::LazyLock;

static VERB_INTENT: LazyLock<HashMap<&str, &str>> = LazyLock::new(|| {
    let mut m = HashMap::new();
    m.insert("get", "query"); m.insert("find", "query"); m.insert("fetch", "query"); m.insert("load", "query"); m.insert("list", "query");
    m.insert("create", "creation"); m.insert("new", "creation"); m.insert("add", "creation"); m.insert("insert", "creation");
    m.insert("update", "mutation"); m.insert("modify", "mutation"); m.insert("set", "mutation");
    m.insert("delete", "deletion"); m.insert("remove", "deletion");
    m.insert("validate", "validation"); m.insert("check", "validation"); m.insert("verify", "validation");
    m.insert("process", "transformation"); m.insert("handle", "transformation"); m.insert("transform", "transformation");
    m.insert("send", "communication"); m.insert("publish", "communication"); m.insert("emit", "communication");
    m.insert("calculate", "calculation"); m.insert("compute", "calculation"); m.insert("count", "calculation");
    m.insert("is", "query"); m.insert("has", "query");
    m
});

/// @docspec:method { since: "3.0.0" }
/// @docspec:intentional "Analyzes function name, signature, and body to extract intent signals across name semantics, guard clauses, branches, error handling, logging, and loop properties"
/// @docspec:deterministic
pub fn extract_from_fn(sig: &syn::Signature, block: &syn::Block) -> Option<Value> {
    let name = sig.ident.to_string();
    if name.starts_with('_') { return None; }

    let words: Vec<&str> = name.split('_').collect();
    let verb = words.first().copied().unwrap_or("");
    let object = words[1..].join("_");
    let intent = VERB_INTENT.get(verb).copied().unwrap_or("unknown");

    let body_text = quote::quote!(#block).to_string();

    // Count patterns in the body text
    let guard_clauses = body_text.matches("return Err").count()
        + body_text.matches("bail!").count()
        + body_text.matches("anyhow!").count();
    let branches = body_text.matches(" if ").count() + body_text.matches("match ").count();
    let error_handling = body_text.matches("Err(").count();
    let log_statements = body_text.matches("log::").count()
        + body_text.matches("tracing::").count()
        + body_text.matches("println!").count()
        + body_text.matches("eprintln!").count();

    let mut signals = serde_json::json!({
        "nameSemantics": {
            "verb": verb,
            "object": object,
            "intent": intent,
        },
    });

    let map = signals.as_object_mut().unwrap();
    if guard_clauses > 0 { map.insert("guardClauses".to_string(), guard_clauses.into()); }
    if branches > 0 { map.insert("branches".to_string(), branches.into()); }
    if error_handling > 0 {
        map.insert("errorHandling".to_string(), serde_json::json!({
            "catchBlocks": error_handling,
            "caughtTypes": [],
        }));
    }
    if log_statements > 0 { map.insert("logStatements".to_string(), log_statements.into()); }

    // Loop properties: detect iterators, for-loops, and stream ops
    let loop_props = extract_loop_properties(&body_text);
    if !loop_props.is_null() {
        map.insert("loopProperties".to_string(), loop_props);
    }

    Some(signals)
}

/// Rust stream/iterator operations to track in `streamOps`.
const STREAM_OPS: &[&str] = &[
    "map", "filter", "fold", "flat_map", "for_each", "collect",
    "any", "all", "find", "position", "enumerate", "zip",
    "take", "skip", "chain", "cloned", "copied",
    "filter_map", "flat_map", "inspect", "peekable",
    "reduce", "scan", "sum", "product", "count",
    "min", "max", "min_by", "max_by", "min_by_key", "max_by_key",
    "partition", "unzip",
];

/// Extract loop and iterator properties from the stringified function body.
///
/// @docspec:deterministic
/// @docspec:intentional "Detects for-loops, iterator chains, and stream operations to populate the loopProperties DSTI channel"
fn extract_loop_properties(body_text: &str) -> Value {
    let has_for = body_text.contains("for ") && body_text.contains(" in ");
    let has_iterators = body_text.contains(".iter()")
        || body_text.contains(".iter_mut()")
        || body_text.contains(".into_iter()");

    // Detect stream ops by looking for `.op_name(` patterns
    let mut stream_ops: Vec<&str> = Vec::new();
    for &op in STREAM_OPS {
        let pattern = format!(".{}(", op);
        if body_text.contains(&pattern) {
            stream_ops.push(op);
        }
    }

    if !has_for && !has_iterators && stream_ops.is_empty() {
        return Value::Null;
    }

    let mut props = serde_json::json!({});
    let obj = props.as_object_mut().unwrap();

    if has_iterators || !stream_ops.is_empty() {
        obj.insert("hasStreams".to_string(), true.into());
    }
    if has_for {
        obj.insert("hasEnhancedFor".to_string(), true.into());
    }
    if !stream_ops.is_empty() {
        stream_ops.sort();
        stream_ops.dedup();
        obj.insert("streamOps".to_string(), serde_json::json!(stream_ops));
    }

    props
}
